File: ade/scheduler.py
"""User-level schedule rendering and installation for ADE integrations."""

from __future__ import annotations

import os
import plistlib
import shlex
import subprocess
import sys
import tempfile
from collections.abc import Mapping
from pathlib import Path
from typing import Any

DEFAULT_TIMES = ("08:00", "12:00", "17:00")
SERVICE_NAME = "ade-plane-linear-sync"
LAUNCHD_LABEL = "com.ade.plane-linear-sync"


class ScheduleError(ValueError):
    """The requested schedule cannot be rendered or installed safely."""


def _validate_times(times: tuple[str, ...] | list[str]) -> tuple[str, ...]:
    value = tuple(times)
    if not value or len(set(value)) != len(value):
        raise ScheduleError("sync schedule must contain unique times")
    for item in value:
        try:
            hour, minute = (int(part) for part in item.split(":", 1))
        except (ValueError, AttributeError):
            raise ScheduleError("sync schedule times must use HH:MM") from None
        if (
            not (0 <= hour <= 23 and 0 <= minute <= 59)
            or len(item) != 5
            or item[2] != ":"
        ):
            raise ScheduleError("sync schedule times must use HH:MM")
    return value


def _command(
    root: Path, provider: str, env_file: Path | None, executable: str | None
) -> list[str]:
    command = [
        executable or sys.executable,
        "-m",
        "ade.cli",
        "--root",
        str(root),
        "sync",
        "--provider",
        provider,
    ]
    if env_file:
        command.extend(["--env-file", str(env_file)])
    return command


def render_systemd(
    root: str | Path,
    provider: str = "plane-linear-sync",
    times: tuple[str, ...] = DEFAULT_TIMES,
    env_file: str | Path | None = None,
    executable: str | None = None,
) -> dict[str, str]:
    times = _validate_times(times)
    root = Path(root).expanduser().resolve()
    env_path = Path(env_file).expanduser().resolve() if env_file else None
    command = shlex.join(_command(root, provider, env_path, executable))
    service = "\n".join(
        [
            "[Unit]",
            "Description=ADE Plane to Linear issue synchronization",
            "",
            "[Service]",
            "Type=oneshot",
            f"ExecStart={command}",
            "",
        ]
    )
    timer = "\n".join(
        [
            "[Unit]",
            "Description=Run ADE Plane to Linear synchronization on schedule",
            "",
            "[Timer]",
            *[f"OnCalendar=*-*-* {time}:00" for time in times],
            "Persistent=true",
            f"Unit={SERVICE_NAME}.service",
            "",
            "[Install]",
            "WantedBy=timers.target",
            "",
        ]
    )
    return {f"{SERVICE_NAME}.service": service, f"{SERVICE_NAME}.timer": timer}


def render_launchd(
    root: str | Path,
    provider: str = "plane-linear-sync",
    times: tuple[str, ...] = DEFAULT_TIMES,
    env_file: str | Path | None = None,
    executable: str | None = None,
) -> bytes:
    times = _validate_times(times)
    root = Path(root).expanduser().resolve()
    env_path = Path(env_file).expanduser().resolve() if env_file else None
    intervals = []
    for item in times:
        hour, minute = (int(part) for part in item.split(":", 1))
        intervals.append({"Hour": hour, "Minute": minute})
    value: dict[str, Any] = {
        "Label": LAUNCHD_LABEL,
        "ProgramArguments": _command(root, provider, env_path, executable),
        "StartCalendarInterval": intervals,
        "RunAtLoad": False,
        "ProcessType": "Background",
        "StandardOutPath": str(root / "state" / provider / "scheduler.stdout.log"),
        "StandardErrorPath": str(root / "state" / provider / "scheduler.stderr.log"),
    }
    return plistlib.dumps(value, fmt=plistlib.FMT_XML, sort_keys=False)


def render(
    platform_name: str,
    root: str | Path,
    provider: str = "plane-linear-sync",
    times: tuple[str, ...] = DEFAULT_TIMES,
    env_file: str | Path | None = None,
    executable: str | None = None,
) -> dict[str, str | bytes]:
    if platform_name == "linux":
        return render_systemd(root, provider, times, env_file, executable)
    if platform_name == "darwin":
        return {
            f"{LAUNCHD_LABEL}.plist": render_launchd(
                root, provider, times, env_file, executable
            )
        }
    raise ScheduleError("ADE scheduling supports Linux systemd and macOS launchd")


def _atomic_write(path: Path, value: str | bytes) -> None:
    path.parent.mkdir(parents=True, exist_ok=True, mode=0o700)
    with tempfile.NamedTemporaryFile(
        "wb", dir=path.parent, prefix=".schedule-", delete=False
    ) as stream:
        temporary = Path(stream.name)
        stream.write(value.encode() if isinstance(value, str) else value)
    temporary.chmod(0o600)
    os.replace(temporary, path)


def install(
    root: str | Path,
    platform_name: str,
    provider: str = "plane-linear-sync",
    times: tuple[str, ...] = DEFAULT_TIMES,
    env_file: str | Path | None = None,
    executable: str | None = None,
    destination: str | Path | None = None,
    enable: bool = False,
) -> dict[str, Any]:
    """Install only ADE-owned user scheduler files.

    ``destination`` is a test/operator override.  Without it the files live in
    the current user's systemd or launchd directory.  Enabling the timer is
    explicit because it changes the user's process manager state.
    """

    root = Path(root).expanduser().resolve()
    rendered = render(platform_name, root, provider, times, env_file, executable)
    state_directory = root / "state" / provider
    state_directory.mkdir(parents=True, exist_ok=True, mode=0o700)
    os.chmod(state_directory, 0o700)
    if destination is not None:
        directory = Path(destination).expanduser().resolve()
    elif platform_name == "linux":
        directory = Path.home() / ".config/systemd/user"
    else:
        directory = Path.home() / "Library/LaunchAgents"
    paths = {}
    for name, content in rendered.items():
        path = directory / name
        _atomic_write(path, content)
        paths[name] = str(path)
    if enable and destination is None and platform_name == "linux":
        subprocess.run(["systemctl", "--user", "daemon-reload"], check=True)
        subprocess.run(
            ["systemctl", "--user", "enable", "--now", f"{SERVICE_NAME}.timer"],
            check=True,
        )
    elif enable and destination is None and platform_name == "darwin":
        subprocess.run(
            [
                "launchctl",
                "bootstrap",
                f"gui/{os.getuid()}",
                paths[f"{LAUNCHD_LABEL}.plist"],
            ],
            check=True,
        )
    return {
        "platform": platform_name,
        "provider": provider,
        "times": list(times),
        "files": paths,
        "enabled": enable,
    }


def manifest_schedule(manifest: Mapping[str, Any]) -> tuple[str, ...]:
    schedule = manifest.get("schedule", {})
    if not isinstance(schedule, Mapping):
        raise ScheduleError("provider schedule must be an object")
    return _validate_times(schedule.get("times", DEFAULT_TIMES))

File: internal/ade/scheduler.go
package ade

import (
	"bytes"
	"encoding/xml"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"runtime"
	"strconv"
	"strings"
)

const defaultScheduleProvider = "plane-linear-sync"
const schedulerService = "ade-plane-linear-sync"
const schedulerLabel = "com.ade.plane-linear-sync"

var defaultSyncTimes = []string{"08:00", "12:00", "17:00"}

func ValidateScheduleTimes(times []string) ([]string, error) {
	if len(times) == 0 {
		return nil, fmt.Errorf("sync schedule must contain unique times")
	}
	seen := map[string]bool{}
	for _, value := range times {
		if !regexp.MustCompile(`^(?:[01][0-9]|2[0-3]):[0-5][0-9]$`).MatchString(value) {
			return nil, fmt.Errorf("sync schedule times must use HH:MM")
		}
		if seen[value] {
			return nil, fmt.Errorf("sync schedule must contain unique times")
		}
		seen[value] = true
	}
	return append([]string(nil), times...), nil
}

func scheduleCommand(root, provider, envFile, executable string) []string {
	if executable == "" {
		executable = os.Getenv("ADE_BIN")
	}
	if executable == "" {
		executable = "/usr/local/bin/ade"
	}
	result := []string{executable, "--root", root, "sync", "--provider", provider}
	if envFile != "" {
		result = append(result, "--env-file", envFile)
	}
	return result
}

func ShellQuote(values []string) string {
	quoted := make([]string, 0, len(values))
	for _, value := range values {
		if value != "" && regexp.MustCompile(`^[A-Za-z0-9_./:=+-]+$`).MatchString(value) {
			quoted = append(quoted, value)
		} else {
			quoted = append(quoted, "'"+strings.ReplaceAll(value, "'", "'\\''")+"'")
		}
	}
	return strings.Join(quoted, " ")
}

func RenderSystemdSchedule(root, provider string, times []string, envFile, executable string) (map[string]string, error) {
	checked, err := ValidateScheduleTimes(times)
	if err != nil {
		return nil, err
	}
	root, err = filepath.Abs(root)
	if err != nil {
		return nil, err
	}
	if envFile != "" {
		envFile, err = filepath.Abs(envFile)
		if err != nil {
			return nil, err
		}
	}
	command := ShellQuote(scheduleCommand(root, provider, envFile, executable))
	service := strings.Join([]string{"[Unit]", "Description=ADE Plane to Linear issue synchronization", "", "[Service]", "Type=oneshot", "ExecStart=" + command, ""}, "\n")
	lines := []string{"[Unit]", "Description=Run ADE Plane to Linear synchronization on schedule", "", "[Timer]"}
	for _, time := range checked {
		lines = append(lines, "OnCalendar=*-*-* "+time+":00")
	}
	lines = append(lines, "Persistent=true", "Unit="+schedulerService+".service", "", "[Install]", "WantedBy=timers.target", "")
	return map[string]string{schedulerService + ".service": service, schedulerService + ".timer": strings.Join(lines, "\n")}, nil
}

func RenderLaunchdSchedule(root, provider string, times []string, envFile, executable string) ([]byte, error) {
	checked, err := ValidateScheduleTimes(times)
	if err != nil {
		return nil, err
	}
	root, err = filepath.Abs(root)
	if err != nil {
		return nil, err
	}
	if envFile != "" {
		envFile, err = filepath.Abs(envFile)
		if err != nil {
			return nil, err
		}
	}
	args := scheduleCommand(root, provider, envFile, executable)
	var b bytes.Buffer
	b.WriteString("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<!DOCTYPE plist PUBLIC \"-//Apple//DTD PLIST 1.0//EN\" \"http://www.apple.com/DTDs/PropertyList-1.0.dtd\">\n<plist version=\"1.0\">\n<dict>\n")
	writeKeyString(&b, "Label", schedulerLabel)
	b.WriteString("<key>ProgramArguments</key><array>\n")
	for _, arg := range args {
		writeString(&b, arg)
	}
	b.WriteString("</array>\n<key>StartCalendarInterval</key><array>\n")
	for _, value := range checked {
		parts := strings.Split(value, ":")
		hour, _ := strconv.Atoi(parts[0])
		minute, _ := strconv.Atoi(parts[1])
		b.WriteString("<dict>\n")
		writeKeyInt(&b, "Hour", hour)
		writeKeyInt(&b, "Minute", minute)
		b.WriteString("</dict>\n")
	}
	b.WriteString("</array>\n")
	b.WriteString("<key>RunAtLoad</key><false/>\n<key>ProcessType</key><string>Background</string>\n")
	writeKeyString(&b, "StandardOutPath", filepath.Join(root, "state", provider, "scheduler.stdout.log"))
	writeKeyString(&b, "StandardErrorPath", filepath.Join(root, "state", provider, "scheduler.stderr.log"))
	b.WriteString("</dict>\n</plist>\n")
	return b.Bytes(), nil
}

func writeString(out *bytes.Buffer, value string) {
	out.WriteString("<string>")
	_ = xml.EscapeText(out, []byte(value))
	out.WriteString("</string>\n")
}
func writeKeyString(out *bytes.Buffer, key, value string) {
	out.WriteString("<key>")
	_ = xml.EscapeText(out, []byte(key))
	out.WriteString("</key>")
	writeString(out, value)
}
func writeKeyInt(out *bytes.Buffer, key string, value int) {
	fmt.Fprintf(out, "<key>%s</key><integer>%d</integer>\n", key, value)
}

func RenderSchedule(platform, root, provider string, times []string, envFile, executable string) (map[string][]byte, error) {
	if platform == "linux" {
		values, err := RenderSystemdSchedule(root, provider, times, envFile, executable)
		if err != nil {
			return nil, err
		}
		result := map[string][]byte{}
		for name, value := range values {
			result[name] = []byte(value)
		}
		return result, nil
	}
	if platform == "darwin" {
		value, err := RenderLaunchdSchedule(root, provider, times, envFile, executable)
		if err != nil {
			return nil, err
		}
		return map[string][]byte{schedulerLabel + ".plist": value}, nil
	}
	return nil, fmt.Errorf("ADE scheduling supports Linux systemd and macOS launchd")
}

func ManifestSchedule(manifest Object) ([]string, error) {
	rawSchedule, exists := manifest["schedule"]
	if !exists {
		return append([]string(nil), defaultSyncTimes...), nil
	}
	schedule, ok := rawSchedule.(map[string]any)
	if !ok {
		return nil, fmt.Errorf("provider schedule must be an object")
	}
	raw, exists := schedule["times"]
	if !exists {
		return append([]string(nil), defaultSyncTimes...), nil
	}
	times, err := validateStringArray(raw, "schedule.times", 1)
	if err != nil {
		return nil, err
	}
	return ValidateScheduleTimes(times)
}

func InstallSchedule(root, platform, provider string, times []string, envFile, executable, destination string, enable bool) (Object, error) {
	root, err := filepath.Abs(root)
	if err != nil {
		return nil, err
	}
	rendered, err := RenderSchedule(platform, root, provider, times, envFile, executable)
	if err != nil {
		return nil, err
	}
	state := filepath.Join(root, "state", provider)
	if err := os.MkdirAll(state, 0o700); err != nil {
		return nil, err
	}
	_ = os.Chmod(state, 0o700)
	var directory string
	if destination != "" {
		directory, err = filepath.Abs(destination)
		if err != nil {
			return nil, err
		}
	} else if platform == "linux" {
		home, err := os.UserHomeDir()
		if err != nil {
			return nil, err
		}
		directory = filepath.Join(home, ".config/systemd/user")
	} else {
		home, err := os.UserHomeDir()
		if err != nil {
			return nil, err
		}
		directory = filepath.Join(home, "Library/LaunchAgents")
	}
	paths := Object{}
	for name, contents := range rendered {
		path := filepath.Join(directory, name)
		if err := atomicFile(path, contents, 0o600); err != nil {
			return nil, err
		}
		paths[name] = path
	}
	if enable && destination == "" && platform == "linux" {
		if err := exec.Command("systemctl", "--user", "daemon-reload").Run(); err != nil {
			return nil, err
		}
		if err := exec.Command("systemctl", "--user", "enable", "--now", schedulerService+".timer").Run(); err != nil {
			return nil, err
		}
	}
	if enable && destination == "" && platform == "darwin" {
		uid := os.Getuid()
		path, _ := paths[schedulerLabel+".plist"].(string)
		if err := exec.Command("launchctl", "bootstrap", fmt.Sprintf("gui/%d", uid), path).Run(); err != nil {
			return nil, err
		}
	}
	return Object{"platform": platform, "provider": provider, "times": times, "files": paths, "enabled": enable}, nil
}

func CurrentPlatform() string { return runtime.GOOS }

func atomicFile(path string, data []byte, mode os.FileMode) error {
	if err := os.MkdirAll(filepath.Dir(path), 0o700); err != nil {
		return err
	}
	file, err := os.CreateTemp(filepath.Dir(path), ".ade-write-")
	if err != nil {
		return err
	}
	temporary := file.Name()
	defer os.Remove(temporary)
	if _, err := file.Write(data); err != nil {
		file.Close()
		return err
	}
	if err := file.Chmod(mode); err != nil {
		file.Close()
		return err
	}
	if err := file.Sync(); err != nil {
		file.Close()
		return err
	}
	if err := file.Close(); err != nil {
		return err
	}
	return os.Rename(temporary, path)
}
